cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Utility methods for decomposing arbitrary n-qubit (2^n x 2^n) unitary."""
+
+from scipy.linalg import cossin
+
+import numpy as np
+
+import cirq
+
+
+def quantum_shannon_decomposition(qubits: list, U: np.ndarray, ops=None):
+    """Returns a list of operations for an arbitrary n-qubit decomposition, preserving phase
+
+    The algorithm is described in Shende et al.:
+    Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+    https://arxiv.org/abs/quant-ph/0406176
+
+    Args:
+        qubits: List of qubits in order of significance
+        U: Numpy array for unitary matrix representing gate to be decomposed
+        ops: List of new existing operations on which to append new operations, whenever
+                a recusive call is made
+             If 'None' is given, a new list is instantiated
+
+    Calls:
+        (Base Case)
+        1._single_qubit_decomposition
+
+            OR
+
+        (Recursive Case)
+        1. _msb_demuxer
+        2. _multiplexed_cossin
+        3. _msb_demuxer
+
+    Returns:
+        List of 2-qubit and 1-qubit operations from the set
+           { CNOT, rz, ry, ZPowGate }
+
+    Raises:
+        ValueError: If the U matrix is not of shape (2^n,2^n)
+        ValueError: If the U matrix is non-unitary
+    """
+    if ops is None:
+        ops = []  # Declare an empty list if no previous operations
+
+    n, m = U.shape[0], U.shape[1]
+    if not (n and not (n & (n - 1)) and n == m):
+        raise ValueError(
+            f"Expected input matrix U to a (2^n x 2^n) shaped numpy array, \
+                but instead got shape {(n,m)}"
+        )
+
+    if not cirq.is_unitary(U):  # Check that U is unitary
+        raise ValueError(
+            "Expected input matrix U to be unitary, \
+                but it fails cirq.is_unitary check"
+        )
```

Comment:
You can move the `is_unitary` check above so you don't need to check that `U` has correct shape again. 
```suggestion
    if not cirq.is_unitary(U):  # Check that U is unitary
        raise ValueError(
            "Expected input matrix U to be unitary, \
                but it fails cirq.is_unitary check"
        )
    n = U.shape[0]
    if n & (n - 1):
        raise ValueError(
            f"Expected input matrix U to be a (2^n x 2^n) shaped numpy array, \
                but instead got shape {U.shape}"
        )
```